File: sdk/typescript/examples/quickstart.mjs
/* ---------------------------------------------------------------------------
   @mcpip/sdk quickstart — the full agent lifecycle against a sandbox gateway.

   Prereq:  MCPIP_SANDBOX_MODE=true uvicorn app.main:app --port 8080
   Run:     node examples/quickstart.mjs        (after: npx tsc -p tsconfig.json)

   Installed usage imports from the package instead:
     import { McpipClient, McpipSandboxClient, McpipDenied, openaiToolCall } from '@mcpip/sdk';
--------------------------------------------------------------------------- */

import {
  McpipClient,
  McpipSandboxClient,
  McpipDenied,
  openaiToolCall,
} from '../dist/index.js';

const BASE = (process.env.MCPIP_BASE ?? 'http://localhost:8080').replace(/\/+$/, '');

// SANDBOX ONLY: /v1/dev/token answers 404 on a production gateway, where your
// own IdP mints the JWT — pass that token (or an async minter hitting your
// STS) as the `token` option instead. devTokenSource re-mints ~30s before
// each short-lived sandbox token expires.
const sandbox = new McpipSandboxClient({ baseUrl: BASE });
const token = sandbox.devTokenSource({
  tenant_id: 'tenant-acme',
  agent_id: 'agent-quickstart',
  role: 'ops',
});

const client = new McpipClient({ baseUrl: BASE, token });
const authenticator = new McpipSandboxClient({ baseUrl: BASE, token });

// 1. What may this identity even name? Metadata only — real targets never
//    cross this boundary.
const catalog = await client.catalog();
console.log('catalog:');
for (const item of catalog) {
  console.log(`  ${item.alias}  [${item.risk_tier}]`);
}

// 2. An auto-tier call: one POST /v1/authorize -> committed receipt.
//    skill_customer_lookup is a default sandbox-catalog row. (Don't pick
//    blindly by tier in the sandbox: some rows are canary decoys, and
//    touching one quarantines the caller — that is the point of them.)
const lookup = await client.authorize({
  source_format: 'openai_tool_call',
  tool_call: openaiToolCall('skill_customer_lookup', { customer_id: 'C-1042' }),
});
if (lookup.status === 'allowed') {
  console.log(`\nallowed: ${lookup.transactionRef} via ${lookup.executedTargetClass}`);
}

// 3. The PIN ceremony on a pin_required alias. The staged result carries the
//    request, and complete() resubmits it verbatim — the payload-bound lock
//    demands an identical consume.
const staged = await client.authorize({
  source_format: 'openai_tool_call',
  tool_call: openaiToolCall('skill_wire_transfer', { amount: '125.00', currency: 'USD' }),
});
if (staged.status === 'staged') {
  console.log(`\nstaged: challenge ${staged.challengeId}`);
  // Out-of-band in production (enrolled authenticator); in the sandbox this
  // endpoint stands in for the device.
  const pin = await authenticator.authenticatorCode(staged.challengeId);
  const committed = await client.complete(staged, pin);
  console.log(`committed after step-up: ${committed.transactionRef}`);
}

// 4. Denials are opaque BY DESIGN: a typed error with a correlation id and
//    the generic message — never a reason. The reason lives in the WORM log,
//    where an operator can look it up by this id.
try {
  await client.authorize({
    source_format: 'openai_tool_call',
    tool_call: openaiToolCall('skill_no_such_alias', {}),
  });
} catch (err) {
  if (err instanceof McpipDenied) {
    console.log(`\ndenied (opaque): "${err.message}" correlation_id=${err.correlationId}`);
  } else {
    throw err;
  }
}

File: sdk/python/examples/quickstart.py
#!/usr/bin/env python3
"""
mcpip-sdk quickstart — the agent surface end to end against a SANDBOX gateway.

Start the gateway first (Redis on :63790, then)::

    MCPIP_SANDBOX_MODE=true uvicorn app.main:app --port 8080

then run::

    python sdk/python/examples/quickstart.py [http://localhost:8080]

Everything below is REAL traffic: a minted sandbox identity, a live catalog
read, an authorized call with its audit receipt, an opaque denial, and the
full PIN step-up ceremony (stage → sandbox authenticator OTP → complete).
"""

from __future__ import annotations

import os
import sys

try:
    from mcpip_sdk import Allowed, MCPIPDenied, SandboxClient, Staged
except ImportError:  # running from the repo without `pip install mcpip-sdk`.
    sys.path.insert(
        0, os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "src")
    )
    from mcpip_sdk import Allowed, MCPIPDenied, SandboxClient, Staged


def main() -> int:
    base_url = sys.argv[1] if len(sys.argv) > 1 else "http://localhost:8080"

    with SandboxClient(base_url) as client:
        # 1) Connectivity — /healthz is unauthenticated and never shed.
        health = client.health()
        print(f"gateway  : {base_url}  v{health.version}  loop={health.loop}")

        # 2) Identity — SANDBOX ONLY. In production your IdP mints the JWT and
        #    you pass it (or a refresh callback) to the client instead.
        client.set_token(lambda: client.dev_token(agent_id="agent-quickstart"))

        # 3) What may this identity even see? Metadata only — never targets.
        items = client.catalog()
        print(f"catalog  : {len(items)} aliases visible")
        for item in items[:5]:
            print(f"           {item.alias}  [{item.risk_tier}]")

        # 4) An AUTO-tier call: authorized, audit-logged, dispatched — one POST.
        outcome = client.authorize("skill_spend_summary", {"period": "2026-Q2"})
        assert isinstance(outcome, Allowed)
        print(f"allowed  : {outcome.transaction_ref}  worm#{outcome.worm_sequence}")

        # 5) A denial is OPAQUE: no reason, only a correlation id an operator
        #    can resolve against the WORM audit log. Never retry a deny.
        try:
            client.authorize("skill_export_everything", {})
        except MCPIPDenied as denied:
            print(f"denied   : correlation_id={denied.correlation_id} (opaque by design)")

        # 6) The PIN ceremony on a pin_required alias: stage → OTP → complete.
        staged = client.authorize("skill_payroll_run", {"run_id": "QS-1"})
        assert isinstance(staged, Staged)
        print(f"staged   : challenge={staged.challenge_id}  ttl={staged.expires_in}s")
        otp = client.authenticator_code(staged.challenge_id)  # sandbox stand-in
        receipt = client.complete(staged, otp)                # identical payload
        print(f"stepped  : {receipt.transaction_ref}  worm#{receipt.worm_sequence}")

    return 0


if __name__ == "__main__":
    raise SystemExit(main())
